out_s3: add support for CSV formatting

edsiper opened the pull request:
The following PR introduces two new configuration properties allowing optional formatting to CSV:

- `format`: defines the data format, it supports `json` and `csv`. Default is `json`.

- `csv_column_names`: when format is `csv`, it appends the column names at the beginning of the S3 object. Default is `off`.

cosmo0920 reviewed (commented):
I'm testing this PR and confirmed that this PR starts to work for CSV formatting.

I added a small nitpick for handling unknown format.

patrick-stephens commented on plugins/out_s3/s3.c:
```diff
@@ -768,6 +768,18 @@ static int cb_s3_init(struct flb_output_instance *ins,
+            ctx->format = S3_FORMAT_CSV;
+        }
+    }
+
```
Can we add a debug statement about what config is used?

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -768,6 +768,18 @@ static int cb_s3_init(struct flb_output_instance *ins,
 
+    /* Formatting */
+    ctx->format = S3_FORMAT_JSON;
+    tmp = flb_output_get_property("format", ins);
```
Looks like this new option is incompatible with our log key option, so that should be called out in the config map help text and guarded against here.

edsiper replied:
@PettitWesley I just pushed a fix to address that

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -2117,17 +2132,35 @@ static void cb_s3_flush(struct flb_event_chunk *event_chunk,
+         */
+        chunk = chunk_tmp;
     }
+
```
we should stop invalid format in the config validation in init? by flush time the config can be assumed to be valid?

cosmo0920 replied:
Yup. We can stop proceeding with invalid format on init. Nothing should do here.
Checking for invalid format should be done before calling this function.

patrick-stephens replied:
Sure, it's just not done at the moment. I tend to be overly defensive due to past experiences of corruption triggering hard to find bugs too!

PettitWesley commented on plugins/out_s3/s3_store.c:
```diff
@@ -179,13 +183,60 @@ int s3_store_buffer_put(struct flb_s3 *ctx, struct s3_file *s3_file,
+            csv_add_columns = FLB_TRUE;
+        }
+
+        if (ctx->log_key) {
+            /*
+             * if log_key option was used, the incoming data is NOT a msgpack object,
+             * likely a JSON string message that we need to convert back so we can
+             * do the CSV formatting.
```
I'm struggling to understand how CSV works with log_key.

So the incoming msgpack is an array of records, each which might have multiple keys. I thought the goal of the CSV column headers is to give names to each of the top level keys in each record? With log key, we just take a single key from each record called `log`. 

In the log_key case, the incoming data is not a JSON. Its a newline delimited buffer of each record's log_key value. That is the value of each log key for each record delimitted by newlines. You can't parse it as JSON unless you do it line by line. Each line might be a JSON or it might just be a string.

pwhelan replied:
As far as I understand it, when `log` key is set it contains the full log in plain text. If that is so we would need to parse it with a fluent-bit parser to transform it into a set of fields.

Perhaps one or several of the following might be a better solution?

- Add a new `log_key_parser` option to use a parser for the log_key.
- Add a new `log_key_format` option to parse it as one of several specific formats.
- Send the `log_key` directly as a line of CSV.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -2117,17 +2132,35 @@ static void cb_s3_flush(struct flb_event_chunk *event_chunk,
+         * for hence the formatting of the CSV will happen inside the function
+         * s3_store_buffer_put().
+         *
+         * For now, we duplicate the buffer for safety reasons.
```
what safety does it bring? I am not understanding why we need to duplicate it here...

pwhelan replied:
I think this comment is also outdated, with respect to the refactoring done to `s3_format`.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -2117,17 +2132,35 @@ static void cb_s3_flush(struct flb_event_chunk *event_chunk,
+         * CSV is a bit tricky, if the user asked to include the column names
+         * we need to know if the content will be placed in a new object or not,
+         * for hence the formatting of the CSV will happen inside the function
+         * s3_store_buffer_put().
```
I think we can determine in this cb_flush whether or not we are creating a new buffer file or not. 


`s3_store_file_get` returns NULL if no existing buffer file exists that we can use. See the `buffer_chunk` call which is the case where we create a new buffer. 

Also, this needs to work with the fact that S3 can upload data without first buffering it. If the new data + any old data is larger than our upload size, we just upload directly. And so we need to handle CSV in that case too. 

I think this should not be implemented in S3 store consequently.

edsiper replied:
@PettitWesley I've refactor the formatter and simplified the logic, now everything is on s3.c

would you please elaborate on the direct upload stuff ?, I got confused with this approach, does it means a previous content with the new incoming one are sent together ? if so, the logic for column headers should work since a file existed previously so no headers were created.

edsiper commented on plugins/out_s3/s3.c:
```diff
@@ -768,6 +768,18 @@ static int cb_s3_init(struct flb_output_instance *ins,
+        }
+        else if (!strcasecmp(tmp, "csv")) {
+            ctx->format = S3_FORMAT_CSV;
+        }
```
fixed.

edsiper commented:
@cosmo0920  @PettitWesley I've refactored the code a little with some cleanups; I've added also formatter callback and proper unit tests. Can you re-check it?

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -2097,6 +2115,103 @@ static void flush_init(void *out_context)
+    else {
+        /*
+         * if 'log_key' is not defined we create a temporary copy of the content since
+         * other parts of the plugin might try to 'release' the buffer which could
+         * generate some issues.
```
I'm still struggling to understand this code... what issues will be generated?

edsiper replied:
I don't remember which exact location happens but that fixed the problem. I wrote couple of unit tests to make sure there are no leaks (just pushed other fixes)

PettitWesley replied:
@edsiper Sorry I need a more definite explanation here, I don't understand why we need to duplicate this memory. Even if there aren't leaks, we still don't want to unnecessarily increase memory usage?

pwhelan replied:
The code in `s3_format` has been refactored to avoid unnecessary copies as much as possible, at least as far as I can tell.

PettitWesley commented on plugins/out_s3/s3.c:
```diff
@@ -2097,6 +2118,91 @@ static void flush_init(void *out_context)
+        }
+
+        if (ctx->format == S3_FORMAT_CSV) {
+            /*
+             * Since log_key option was used, the incoming data is NOT a msgpack object,
+             * likely a JSON string message that we need to convert back so we can do the CSV formatting.
+             */
+
+            ret = flb_pack_json(data, bytes, &mp_buf, &mp_size, &root_type);
+            if (ret < 0) {
+                return NULL;
+            }
```
As I commented before, how does this work? With log key, the value of the log_key is used instead of the whole record. But that's for each record. So the buffer as a whole would be a newline delimited list of the value of each log_key. 

So it might be:
```
val1
val2
val3
val4
```

Or it could be a series of JSONs:

```
{ "embeddedjson": "some value which was inside the log key1"}
{ "embeddedjson": "some value which was inside the log key2"
{ "embeddedjson": "some value which was inside the log key3"
```

In either case, you can not treat the buffer as if it was just a single JSON.

pwhelan replied:
I have also been thinking about the same problem. Several ideas I have had are:

- Add support for using parsers on the `log_key`.
- Add a new property to specify a format to parse the `log_key` with.
- Throw an error or warning when using CSV output format without any specified way to parse the `log_key`.

pwhelan replied:
Any thoughts, @edsiper or @PettitWesley?

PettitWesley replied:
I think that to move forward with this CSV feature quickly we should just have it not apply to the log_key option. 

As a longer term enhancement, adding support for a parser for the log_key sounds like an interesting idea. How would that work? Use CSV parser to individually convert each log_key value to a CSV?
